File: exercises/00 my first unit test/java/src/eu/sioux/tdd/myfirst/MyDateMatcher.java
package eu.sioux.tdd.myfirst;

import java.util.ArrayList;
import java.util.Date;
import java.util.Iterator;
import java.util.List;

public class MyDateMatcher {
    private final List<Date> myDates;

    public MyDateMatcher(List<Date> dates) {
        myDates = dates;
    }

    public List<Date> DatesAreMatching(FuzzyDate fuzzyDate) {
        List<Date> matches = new ArrayList<>();
        matches.clear();

        Iterator<Date> it = myDates.iterator();

        while (it.hasNext()) {
            Date date = (Date) it.next();
            if (fuzzyDate.contains(date)) {
                matches.add(date);
            }
        }

        return matches;
    }
}

File: solutions/00 my first unit test/cpp/My first unit test/My first unit test/MyDateMatcher.cpp
#include "MyDateMatcher.h"

#include <algorithm>
#include <iterator>
#include "FuzzyDate.h"

using namespace std;
using namespace date;

MyDateMatcher::MyDateMatcher(const vector<year_month_day>& dates) :
	_myDates(dates)
{
}

MyDateMatcher::~MyDateMatcher()
{
}

vector<year_month_day> MyDateMatcher::FindMatchingDates(const FuzzyDate& fuzzyDate) const
{
	vector<year_month_day> matchingDates;

	copy_if(_myDates.begin(), _myDates.end(), back_inserter(matchingDates), [fuzzyDate](const year_month_day& date) {return fuzzyDate.contains(date);});

	return matchingDates;
}
